### PR TITLE
ZCP List Bookmarks & Holds

### PR DESCRIPTION
**Pull request by sdimitro**
# Commit

DLPX-50618 ZFS channel programs - get holds
DLPX-50362 ZFS channel programs - list bookmarks
DLPX-52112 rename zfs.list.properties to zfs.list.user_properties

# Testing Done

The channel programs from the test suite pass (including the new ones).
The only exception is LX-1341.

I also ran the integration tests that depend on these changes and the pass. Output:
```
====================================================================
UNEXPECTED TEST SUCCESSES FOR MODULE :misc:zfs-data
====================================================================
  com.delphix.misc.zfsdata.test.MdsZfsUtilsTest.liveAndConsistentMatchForTestDomainBookmarksAndHolds
  com.delphix.misc.zfsdata.test.MdsZfsUtilsTest.liveAndConsistentMatchForGetSnapshots
  com.delphix.misc.zfsdata.test.MdsZfsUtilsTest.doInitWithExistingCloneAndSnapshot
  com.delphix.misc.zfsdata.test.MdsZfsUtilsTest.consistentMatchesExpectedForTestDomain
  com.delphix.misc.zfsdata.test.MdsZfsUtilsTest.doInit
  com.delphix.misc.zfsdata.test.MdsZfsUtilsTest.liveAndConsistentMatchForTestDomain
  com.delphix.misc.zfsdata.test.MdsZfsUtilsTest.checkKeyFormatInChildrenAndSnapshotsMap
  com.delphix.misc.zfsdata.test.MdsZfsUtilsTest.liveAndConsistentMatchForGetSupportDatasets
  com.delphix.misc.zfsdata.test.MdsZfsUtilsTest.liveAndConsistentMatchForDatasetExists
====================================================================
```

### CODE REVIEW

**Comment by ahrens** [comment]
Can you list which DelphixOS commits this corresponds to?  I think that will help us track what's in which repos.

File: module/zfs/zcp_iter.c [hunk 1]
```diff
@@ -361,11 +360,11 @@ zcp_props_iter(lua_State *state)
...
 }
 
-static int zcp_props_list(lua_State *);
-static zcp_list_info_t zcp_props_list_info = {
-	.name = "properties",
-	.func = zcp_props_list,
-	.gc = zcp_props_list_gc,
+static int zcp_user_props_list(lua_State *);
+static zcp_list_info_t zcp_user_props_list_info = {
+	.name = "user_properties",
```
[the remark this answers is not in the document]
Comment (sdimitro) [reply module/zfs/zcp_iter.c]: Added JIRA-IDs to description and commit message.
As for changing the name `"properties"`, Tony Hutter stated that the original changes are not in 0.7.X.X so it should be ok to change this.

File: module/zfs/zcp_iter.c [hunk 2]
```diff
@@ -468,6 +468,213 @@ zcp_system_props_list(lua_State *state)
...
+		    err));
+	}
+
+	if (dsl_dataset_phys(ds)->ds_userrefs_obj == 0) {
+		dsl_dataset_rele(ds, FTAG);
+		return (0);
+	}
+
+	zap_cursor_init_serialized(&zc, ds->ds_dir->dd_pool->dp_meta_objset,
+	dsl_dataset_phys(ds)->ds_userrefs_obj, cursor);
```
Comment (jgallag88) [inline module/zfs/zcp_iter.c]: [nit] Indentation

File: module/zfs/zcp_iter.c [hunk 3]
```diff
@@ -468,6 +468,213 @@ zcp_system_props_list(lua_State *state)
...
+	if (ds->ds_bookmarks_obj == 0) {
+		dsl_dataset_rele(ds, FTAG);
+		return (0);
+	}
+
+	/* Store the dataset's name so we can append the bookmark's name */
+	dsl_dataset_name(ds, ds_name);
+
+	zap_cursor_init_serialized(&zc, ds->ds_dir->dd_pool->dp_meta_objset,
+	ds->ds_bookmarks_obj, cursor);
```
Comment (jgallag88) [inline module/zfs/zcp_iter.c]: [nit] Should this continuation be indented further? It might just be the way github is rendering the diff.

Comment (sdimitro) [reply module/zfs/zcp_iter.c]: No you are right, github is rendering those fine

**Comment by sdimitro** [comment]
I ran checkstyle and the only errors were:
```
Running checkstyle
error: commit subject over 50 characters
error: missing "Signed-off-by"
Makefile:1517: recipe for target 'commitcheck' failed
```
 which should be fine for our case

**Pull request merged**